File: Data Structures & Algorithms/two-integer-sum/submission-3.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer,Integer> complement = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            int complementNum = target - num;
            if (complement.containsKey(num)) {
                return new int[] {complement.get(num), i};
            }
            complement.put(complementNum, i);
        }
        return new int[]{-1, -1};
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-3.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i, num in enumerate(nums):
            if num > 0:
                break
            if i > 0 and num == nums[i - 1]:
                continue 
            k = i
            l = i + 1
            r = len(nums) - 1
            while (l < r):
                if nums[l] + nums[r] + num == 0:
                    res.append([nums[l],nums[r],nums[k]])
                    l += 1
                    r -= 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1
                elif nums[l] + nums[r] + num < 0:
                    l += 1
                else:
                    r -= 1
        return res